### PR TITLE
Handle Headless Services when no ports are present

### PR DESCRIPTION
**Pull request by sebgoa**
Addresses #146

### CODE REVIEW

**Comment by kadel** [comment]
When I tried that I got error:
```
time="2016-09-19T13:47:10+02:00" level=warning msg="[redis] No ports defined, we will create a Headless service." 
error validating "STDIN": error validating data: found invalid field portalIP for v1.ServiceSpec; if you choose to ignore these errors, turn validation off with --validate=false
```
When I removed `portalIP` from generated service everything worked fine.

**Comment by sebgoa** [comment]
what compose file did you give it ?

**Comment by kadel** [comment]
```
version: "2"

services:
    web:
      image: tuna/docker-counter23
      ports:
        - "5000:5000"
      links:
        - redis
      networks:
        - default
    
    redis:
      image: redis:3.0
      networks: 
        - default

```

**Comment by kadel** [comment]
I've done some digging around to figure out where PortalIP is coming from, because PortalIP is old deprecated name for ClusterIP.

Because of OpenShift we are using  Kubernetes  from [openshift/kubernetes](https://github.com/openshift/kubernetes).
Problem is that they added [this](https://github.com/openshift/kubernetes/blob/57fb9acc109285378ecd0af925c8160eb8ca19e6/pkg/api/v1/conversion.go#L618) for backward compatibility :-(
Result of this is that Services with PortalIP can be deployed to OpenShift but it fails on Kubernetes :-(

We need to figure the way  how to use upstream Kubernetes for Kubernetes conversion. But I don't have idea how to do it :-( OpenShift is not doing import rewrite so it includes k8s.io/kubernetes but expect this to be their fork of k8s from  [openshift/kubernetes](https://github.com/openshift/kubernetes) :-(

File: pkg/transformer/kubernetes/k8sutils.go [hunk 1]
```diff
@@ -246,6 +246,27 @@ func CreateService(name string, service kobject.ServiceConfig, objects []runtime
 	return svc
 }
 
+// create a k8s headless service
+func CreateHeadlessService(name string, service kobject.ServiceConfig, objects []runtime.Object) *api.Service {
+	svc := InitSvc(name, service)
+
+    servicePorts := []api.ServicePort{}
```
Comment (cdrage) [inline pkg/transformer/kubernetes/k8sutils.go]: i believe a tab is missing here (needs to be ran through gofmt)

**Comment by kadel** [comment]
I just tested this with newer openshift version and PortalIP is gone.
OpenShift and K8S upgrade is part #200.  As soon as #200 get merged we can merge this.

**Comment by surajssd** [comment]
And #200 is merged!

**Comment by sebgoa** [comment]
and we are stuck on my CLA, even though I can probably override this...

**Comment by sebgoa** [comment]
I will rebase tomorrow anyway

File: pkg/transformer/kubernetes/k8sutils.go (hunk 1 above)
Comment (sebgoa) [inline pkg/transformer/kubernetes/k8sutils.go]: fixing in next push, thanks

**Comment by ngtuna** [comment]
@sebgoa could you check the failed test-cases ? That would be great to introduce handless svc support at kubecon talk.

**Comment by sebgoa** [comment]
@kadel @surajssd you want to have a look at this ? I fixed the headless stuff, it is rebased...

**Comment by sebgoa** [comment]
ok tests fixed

**Comment by kadel** [comment]
I've rebased this and added headless services for OpenShift.

ping @sebgoa @ngtuna

**Comment by ngtuna** [comment]
👍 @kadel Could you add an unit test for this ?

**Comment by kadel** [comment]
Of course

File: pkg/app/app_test.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package app
+
+import "testing"
+
+func TestCompareUpAndConvert(t *testing.T) {
+}
```
Comment (cdrage) [inline pkg/app/app_test.go]: blank test, wut

File: pkg/transformer/kubernetes/k8sutils.go [hunk 3]
```diff
@@ -258,7 +258,7 @@ func convertToVersion(obj runtime.Object, groupVersion unversioned.GroupVersion)
 // PortsExist checks if service has ports defined
 func (k *Kubernetes) PortsExist(name string, service kobject.ServiceConfig) bool {
 	if len(service.Port) == 0 {
-		logrus.Warningf("[%s] Service cannot be created because of missing port.", name)
+		logrus.Warningf("[%s] No ports defined, we will create a Headless service.", name)
```
Comment (cdrage) [inline pkg/transformer/kubernetes/k8sutils.go]: Still need to be fixed :( I think the log can be more explicit.

File: pkg/transformer/kubernetes/k8sutils.go [hunk 4]
```diff
@@ -282,6 +282,27 @@ func (k *Kubernetes) CreateService(name string, service kobject.ServiceConfig, o
 	return svc
 }
 
+// CreateHeadlessService creates a k8s headless service
+func (k *Kubernetes) CreateHeadlessService(name string, service kobject.ServiceConfig, objects []runtime.Object) *api.Service {
```
Comment (cdrage) [inline pkg/transformer/kubernetes/k8sutils.go]: What is a 'Headless' service? *I* know what it means, but perhaps adding a comment to the kubernetes documentation may help.

File: pkg/transformer/kubernetes/k8sutils_test.go [hunk 5]
```diff
@@ -161,3 +163,22 @@ func TestIsDir(t *testing.T) {
...
+		ServiceConfigs: map[string]kobject.ServiceConfig{"app": service},
+	}
+	k := Kubernetes{}
+
+	objects := k.Transform(komposeObject, kobject.ConvertOptions{CreateD: true, Replicas: 1})
+	if err := testutils.CheckForHeadless(objects); err != nil {
+		t.Error(err)
+	}
+
+}
```
Comment (cdrage) [inline pkg/transformer/kubernetes/k8sutils_test.go]: :+1:

File: pkg/transformer/kubernetes/k8sutils.go [hunk 6]
```diff
@@ -282,6 +282,27 @@ func (k *Kubernetes) CreateService(name string, service kobject.ServiceConfig, o
...
+
+	svc.Spec.Ports = servicePorts
+	svc.Spec.ClusterIP = "None"
+
+	// Configure annotations
+	annotations := transformer.ConfigAnnotations(service)
+	svc.ObjectMeta.Annotations = annotations
+
+	return svc
+}
```
Comment (cdrage) [inline pkg/transformer/kubernetes/k8sutils.go]: :-1:

File: pkg/app/app_test.go (hunk 2 above)
Comment (kadel) [inline pkg/app/app_test.go]: removing :-) that mistake, you were faster than me :-)

File: pkg/transformer/kubernetes/k8sutils.go (hunk 6 above)
Comment (kadel) [inline pkg/transformer/kubernetes/k8sutils.go]: :question:

Comment (cdrage) [inline pkg/transformer/kubernetes/k8sutils.go]: woops, that was suppose to be a :+1:

File: pkg/transformer/kubernetes/k8sutils.go (hunk 4 above)
Comment (kadel) [inline pkg/transformer/kubernetes/k8sutils.go]: Update, I've tried explain why we use Headless services in comment

File: pkg/transformer/kubernetes/k8sutils.go (hunk 3 above)
Comment (kadel) [inline pkg/transformer/kubernetes/k8sutils.go]: I know :-( But I don't have anything short, that might be more helpful :-(

File: pkg/transformer/kubernetes/k8sutils.go (hunk 6 above)
Comment (kadel) [inline pkg/transformer/kubernetes/k8sutils.go]: ok :wink:

File: pkg/transformer/kubernetes/k8sutils.go (hunk 3 above)
Comment (kadel) [inline pkg/transformer/kubernetes/k8sutils.go]: Does user even cares about this?
Maybe we can get rid of this message completely.

Comment (cdrage) [inline pkg/transformer/kubernetes/k8sutils.go]: @kadel I agree. Removing the log would be best, or adding it to Debug instead. If a port isn't passed, it's assumed to be headless anyways.

Comment (kadel) [inline pkg/transformer/kubernetes/k8sutils.go]: yes, 
changed to debug

**Pull request merged**